crates/matrix-sdk/src/encryption/secret_storage.rs

```diff
@@ -0,0 +1,484 @@
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use ruma::events::secret::request::SecretName;
+    ///
+    /// let secret_store = client
+    ///     .encryption()
+    ///     .open_secret_store("It's a secret to everybody")
+    ///     .await?;
+    ///
+    /// let my_secret_name = SecretName::from("m.treasure");
+    ///
+    /// let secret = secret_store.get_secret(&my_secret_name).await?;
+    ///
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub async fn get_secret(&self, secret_name: &SecretName) -> Result<Option<String>> {
+        let event_type = GlobalAccountDataEventType::from(secret_name.as_str().to_owned());
+
+        if let Some(secret_content) = self.client.account().fetch_account_data(event_type).await? {
+            let mut secret_content = secret_content.deserialize_as::<SecretEventContent>()?;
+
+            if let Some(secret_content) = secret_content.encrypted.remove(self.key.key_id()) {
+                let decrypted = self
+                    .key
+                    .decrypt(&secret_content.try_into()?, secret_name)
+                    .map_err(DecryptionError::from)?;
+                let secret = String::from_utf8(decrypted).map_err(DecryptionError::from)?;
+
+                Ok(Some(secret))
+            } else {
+                Ok(None)
+            }
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Store a secret in the homeserver's account data
+    ///
+    /// This method allows you to securely store a secret on the Matrix
+    /// homeserver as an encrypted account data event.
+    ///
+    /// # Arguments
+    ///
+    /// - `secret_name`: The name of the secret. The provided `secret_name`
+    ///   serves as the event type for the account data event on the homeserver.
+    ///
+    /// - `secret`: The secret to be stored on the homeserver. The secret is
+    ///   encrypted before being stored, ensuring its confidentiality and
+    ///   integrity.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::Client;
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use ruma::events::secret::request::SecretName;
+    ///
+    /// let secret_store = client
+    ///     .encryption()
+    ///     .open_secret_store("It's a secret to everybody")
+    ///     .await?;
+    ///
+    /// let my_secret = "Top secret secret";
+    /// let my_secret_name = SecretName::from("m.treasure");
+    ///
+    /// secret_store.store_secret(&my_secret_name, my_secret);
+    ///
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub async fn store_secret(&self, secret_name: &SecretName, secret: &str) -> Result<()> {
+        let event_type = GlobalAccountDataEventType::from(secret_name.as_str().to_owned());
```

Comment:
I reckon `SecretName` could implement a `to_global_account_data_event_type()` method, as this `GlobalAccountDataEventType::from(…)` dance is happening more than once. Or simply, just a `From<_>` implementation. Thoughts?

Reply:
Hmm, a `From<_>` implementation in Ruma might make sense. I'll open a PR there.

Reply:
https://github.com/ruma/ruma/pull/1668